src/ruleset_info.hpp

```diff
@@ -240,7 +182,28 @@ class ruleset_info : public base_ruleset_info {
         }
     }
 
-    void set_error(std::string error) override { error_ = std::move(error); }
+    void set_error(std::string error) { error_ = std::move(error); }
+
+    [[nodiscard]] ruleset_info_state state() const noexcept
+    {
+        if (!error_.empty()) {
+            return ruleset_info_state::invalid;
+        }
+
+        auto final_state = ruleset_info_state::indeterminate;
```

Comment:
not sure indeterminate passes the right message. It seems like the object is an indeterminate state, but this seems like a valid state. Maybe "empty"?